### PR TITLE
Log referrer in telemetry on Sourcegraph Cloud

### PR DESCRIPTION
**Pull request by dadlerj**
From discussions with the product team, it sounds like seeing each session's initial referral source URL would be useful. Find more context in Slack at https://sourcegraph.slack.com/archives/C027H9ATJAC/p1625664484002900.

I followed the implementation of the `FirstSourceURL` parameter, as it works in a similar way—it's only logged on Sourcegraph.com, not for any private or self-hosted instances.

### CODE REVIEW

**Review by attfarhan — APPROVED** [review]
LGTM

File: client/web/src/user/settings/backend.tsx [hunk 1]
```diff
@@ -135,6 +135,7 @@ export function logEvent(event: string, eventProperties?: unknown): void {
                 $userCookieID: String!
                 $cohortID: String
                 $firstSourceURL: String!
+                $referrer: String!
```
Comment (eseliger) [inline client/web/src/user/settings/backend.tsx:138]:
```suggestion
                $referrer: String
```

File: client/web/src/tracking/eventLogger.ts [hunk 2]
```diff
@@ -113,6 +114,21 @@ export class EventLogger implements TelemetryService {
         return firstSourceURL
     }
 
+    public getReferrer(): string {
+        const referrer = document.referrer
+        try {
+            // 🚨 SECURITY: Return if the referrer is a valid non-Sourcegraph.com URL
+            // to avoid leaking private code URLs.
+            const url = new URL(referrer)
+            if (url.hostname !== 'sourcegraph.com') {
```
Comment (eseliger) [inline client/web/src/tracking/eventLogger.ts:123]: Isn't this the wrong way around? Currently it only returns referrers if the source is not sourcegraph cloud 

```suggestion
            if (url.hostname === 'sourcegraph.com') {
```

**Review by eseliger — APPROVED** [review]

File: client/web/src/tracking/eventLogger.ts (hunk 2 above)
Comment (attfarhan) [reply client/web/src/tracking/eventLogger.ts:123]: That's correct! We want to know the URL a user comes from if they're coming from outside Sourcegraph Cloud. The use case is for product to know which sites (GitHub, Google, etc.) a user comes from before landing on Sourcegraph and whether that changes use behavior. The reason we don't want to track internal URLs is that if they switch pages from a private Sourcegraph Cloud repo to another page, we don't want to send that private URL back to our data stores.

Comment (eseliger) [reply client/web/src/tracking/eventLogger.ts:123]: Discussed via slack, this is not to disable it on premise but to hide private repo names.

**Comment by malomarrec** [comment]
@attfarhan
**Questions**
- How will we distinguish a user that came from an internal page from a user that landed from directly typing in a sourcegraph URL?
- How will we know if a user comes from a docs.sourcegraph.com or about.sourcegraph.com property?

Some suggestions to make this usable in Amplitude (happy to add to a separate ticket):
In Amplitude what I want to do is:
- filter by source, and measure retention and user success
- filter by (google) search keyword if provided, and understand how we're doing

It would be useful to have:
- a `referrer` property with the full URL of the 
- a `referrer_type` user property with (search, direct, referral campaign (for URLS with marketing tags))
- a `search_keyword` with the google search keyword if provided

The `referrer_type` is the most urgent to have. The others are complement for drill-downs.

**Comment by malomarrec** [comment]
@attfarhan
**Questions**
- How will we distinguish a user that came from an internal page from a user that landed from directly typing in a sourcegraph URL?
- How will we know if a user comes from a docs.sourcegraph.com or about.sourcegraph.com property?

Some suggestions to make this usable in Amplitude (happy to add to a separate ticket):
In Amplitude what I want to do is:
- filter by source, and measure retention and user success
- filter by (google) search keyword if provided, and understand how we're doing

It would be useful to have:
- a `referrer` property with the full URL of the 
- a `referrer_type` user property with (search, direct, referral campaign (for URLS with marketing tags))
- a `search_keyword` with the google search keyword if provided

The `referrer_type` is the most urgent to have. The others are complement for drill-downs.

**Comment by attfarhan** [comment]
> How will we distinguish a user that came from an internal page from a user that landed from directly typing in a sourcegraph URL?

I can make it so that if it's a sourcegraph.com hostname, we just log the hostname portion (https://sourcegraph.com) I'll open that in a separate PR.

> How will we know if a user comes from a docs.sourcegraph.com or about.sourcegraph.com property?
We still get about.sourcegraph.com and sourcegraph.com referrers

> Some suggestions to make this usable in Amplitude (happy to add to a separate ticket):
In Amplitude what I want to do is:


To summarize your asks:



> a referrer property with the full URL of the 
This is done


> a referrer_type user property with (search, direct, referral campaign (for URLS with marketing tags))

This is needed for this, right? You can do this with `referrer`, but you want to segment by type

> filter by source, and measure retention and user success

For this, we'd need to implement these guidelines: https://knowledge.hubspot.com/reports/understand-hubspots-traffic-sources-in-the-traffic-analytics-tool#definitions-and-drill-down-data-of-each-source.

Mind creating a separate ticket for this, and let me know what's the deadline?

> a search_keyword with the google search keyword if provided

I have a ticket for this here https://github.com/sourcegraph/analytics/issues/231

**Pull request merged**